src/realm/object-store/sectioned_results.cpp

```diff
@@ -0,0 +1,405 @@
+    key = **ptr;
+    return ptr;
+}
+
+// This method will run in the following scenarios:
+// - SectionedResults is performing its initial evaluation.
+// - The underlying Table in the Results collection has changed
+void SectionedResults::calculate_sections()
+{
+    m_prev_sections.clear();
+    m_prev_section_index_to_key.clear();
+    for (auto& s : m_sections) {
+        m_prev_sections[s.key] = s.index;
+        m_prev_section_index_to_key[s.index] = s.key;
+    }
+
+    bool uses_buffer = false;
+    size_t next_buffer_size = 0;
+    size_t size = m_results.size();
+    std::map<Mixed, size_t> key_to_section_index;
+
+    m_sections.clear();
+    m_section_key_to_index_lookup.clear();
+    m_row_to_index_path.clear();
+    m_row_to_index_path.resize(size);
+
+    for (size_t i = 0; i < size; ++i) {
+        Mixed key = m_callback(m_results.get_any(i), m_results.get_realm());
+        // Disallow links as section keys. It would be uncommon to use them to begin with
+        // and if the object acting as the key was deleted bad things would happen.
+        if (!key.is_null() && key.get_type() == type_Link) {
+            throw std::logic_error("Links are not supported as section keys.");
+        }
+
+        auto it = key_to_section_index.find(key);
+        if (it == key_to_section_index.end()) {
+            util::Optional<std::unique_ptr<util::Optional<std::string>>> buffered_key;
+            if (!key.is_null() && (key.get_type() == type_String || key.get_type() == type_Binary)) {
+                auto ptr = (key.get_type() == type_String) ? create_buffered_key(key, key.get_string())
+                                                           : create_buffered_key(key, key.get_binary());
+                buffered_key = std::move(ptr);
+                uses_buffer = true;
+            }
+
+            auto idx = m_sections.size();
+            Section section;
+            section.key = key;
+            section.index = idx;
+            section.indices.push_back(i);
+            m_sections.push_back(section);
+            it = key_to_section_index.insert({key, m_sections.size() - 1}).first;
+            m_row_to_index_path[i] = {it->second, section.indices.size() - 1};
+            m_section_key_to_index_lookup[key] = idx;
+
+            if (uses_buffer && buffered_key) {
+                m_str_data_buffers.push_back(std::move(*buffered_key));
+                next_buffer_size++;
+            }
+        }
+        else {
+            auto& section = m_sections[it->second];
+            section.indices.push_back(i);
+            m_row_to_index_path[i] = {it->second, section.indices.size() - 1};
+        }
+    }
+    if (uses_buffer) {
+        auto buffers_to_purge = m_str_data_buffers.size() - (m_prev_str_data_buffer_size + next_buffer_size);
+        m_str_data_buffers.erase(m_str_data_buffers.begin(), m_str_data_buffers.begin() + buffers_to_purge);
+        m_prev_str_data_buffer_size = next_buffer_size;
+    }
```

Comment:
I have made it so that the buffers are stored in a `unordered_map<std::shared_ptr<std::string>>`. The reason for this is that if there are multiple change notifications handlers for a `ResultsSection` on a `SectionedResults`, the first change listener will run `calculate_results` and during the `.clear()` of the previous buffers the required buffer for filtering on the section will be lost. To prevent that from happening the `SectionedResultsNotificationHandler` will take a reference to the shared pointer stored for the key so that the only keys are purged from memory are the ones that are not used by any other function.